### ISSUE TITLE
Multithreaded job/task system

### ISSUE THREAD

**Issue by slime73** [love2d/love#2020]
Computers, consoles, and phones all have more than one CPU core these days. But aside from audio/video decoding, all of love's own internal code is singlethreaded. We do have APIs to run Lua code in multiple OS threads, but because each thread needs its own Lua instance it takes work to use efficiently in a lot of situations - independent heavy algorithm processing and asset loading are the two easiest ways to use love's threads right now.

Most modern game engines have an integrated job/task system they use with a thread pool, to parallelize their own work (for example render command generation or physics simulation) and often to parallelize game code, all in a fairly simple manner.

It'd be nice to have something like that in love - maybe just for internal use to start (if good uses can be found) since more infrastructure and profiling work would need to be done to make it really solid in Lua code. In particular it becomes a lot easier to know how to use and misuse a job/task system when there's a way to visualize the different parts of a frame and what job threads are doing in a given section, and there would be details to figure out for keeping jobified Lua functions side effect free in a given job thread.

[enkiTS](https://github.com/dougbinks/enkiTS) seems like a solid library for the base of this sort of thing.

**Comment by HDPLocust** [comment]
It is also very good to transfer network interaction (especially UDP) to a child thread, otherwise lags in the game can lead to the loss of datagrams, both receiving and sending. Perhaps it would also be nice to transfer interaction with ENet to other threads.

**Comment by HDPLocust** [comment]
> However I'm not sure even a dedicated thread is super useful most of the time for realtime networking.

Yes, no one is stopping you from using love2d threads for network activity in versions 7-12. There is no need to build them into the framework.

**Comment by bjornbytes** [comment]
I've been designing a task system for LÖVR, and just opened a PR with an early version of the implementation and notes about the Lua API:

https://github.com/bjornbytes/lovr/pull/916

I think this kind of task system could work in LÖVE as well, if there's interest.  I'm curious if any LÖVE people have thought about this kind of thing already and have ideas of their own, or feedback on this one.

**Comment by HDPLocust** [comment]
> I think this kind of task system could work in LÖVE as well, if there's interest. I'm curious if any LÖVE people have thought about this kind of thing already and have ideas of their own, or feedback on this one.

I made some simple libs for such stuff. All with "pure Lua".

Simple coroutine based promise-like thing
https://gist.github.com/HDPLocust/306396e7ac24d7a99cafb87afd4e3116
Returns 6 args max to not generate tables on each check "did corutine return anything?".

```lua
local function foo(self, x, y, a, b)
	print(os.clock(), "Start 1", x, y, a, b)
	local result = a + b
	self.sleep(.2)

	local foo, err = self.await(
		function(self, a, b)
			print(os.clock(), "Start 2", a, b)
			
			local result = a - b
			print(os.clock(), "Process 2", result)
			self.sleep(2) -- yields until done too
			return result
		end,
		a, b
	)
	
	print("foo", foo)
	return (foo or 10) + result
end

local coro = async(
	foo, 10, 20, 30, 40
):next(
	function(self, x)
		print(os.clock(), "Start next 1", x)
		return x
	end
):next(
	function(self, x)
		print(os.clock(), "Start next 2", x)
	end
)

while async.getTaskCount() > 0 do
	async.update()
end
```

I'm using it like this:
https://github.com/user-attachments/assets/4c9404d7-66bb-48a2-834e-907e62a9f15b



Love2d thread task-queue thing with callbacks like `onsuccess = function() ... end`
https://gist.github.com/HDPLocust/bf5e63e6730b39bbd8a11fc5a0a99847


```lua
local gettime = love.timer.getTime()
local pool = tasker:new(5) -- create pool of 5 threads to deal with tasks

print(gettime(), "Add task")

-- there is single limitation due to string.dump will not serialize upvalues
-- "require" everything you need for this function inside it or pass as arguments
local function HttpRequest(url)
	local http  = require'socket.http'
	local ltn12 = require'socket.ltn12'
	local respbody = {}
	local _, status, headers = http.request{
		url  = url, 
		sink = ltn12.sink.table(respbody)
	}
	return table.concat(respbody), status, headers
end

pool:addTask{
	func = HttpRequest,
	args = {"http://google.com"},
	onsuccess = function(body, status, headers) 
		print(gettime(), "Body:", body, "\nStatus", status) 
	end,
	onerror   = function(err) print("Error!", err) end
}

function love.update(dt)
	print(gettime(), "Call update")
	pool:update()
end
```

But. Everything async/multithreading stuff requires some understanding of "what is going on" from user, so it may lead to unnoticed overcomplication.
